### PR TITLE
Add ability to "reserve" devices by player in Android using settings

### PR DESCRIPTION
**Pull request by PatrickStankard**
## Description

I use RetroArch on an [Arcade1Up Simpsons arcade cabinet](https://arcade1up.com/products/the-simpsons-arcade-cabinet) running Android 10. I wanted to be able to have a controller always map to a specific player when it's plugged in, regardless of the order that it was initialized by RetroArch.

This PR adds settings to the configuration, so that you can "reserve" a device for a given player, based on it's VID/PID, or name. I based this on how the `input_android_physical_keyboard` setting is handled. This is very useful in my specific situation (running a MAME 2003-plus core where the player order needs to match the order of the controllers mounted in the cabinet), but I could image it being useful in other situations, too.

Right now, only the Android input driver is checking to see if a port is reserved, based on the config value. However, other input drivers utilize this change by calling `input_device_get_reserved_port`.

## Related Issues

#12924 

## Reviewers

I'm not sure who to tag for a review, but I'll post the PR in Discord

### CODE REVIEW

**Comment by hizzlekizzle** [comment]
let's see if @sonninnos and/or @LibretroAdmin care to take a look

**Comment by PatrickStankard** [comment]
In this commit (https://github.com/libretro/RetroArch/pull/16269/commits/e148c31b0267b40400c77bca3b355d9cb23873ee), I made it so that the settings themselves, and the method to get a reserved port based on the settings, was no longer Android specific. However, the only place that it's being used is in the Android input driver: https://github.com/libretro/RetroArch/blob/928b1e3c581cae652ea5589ad1e043eaef975e1b/input/drivers/android_input.c#L1367-L1372

So it only works on Android, but it makes it so that other platforms and input drivers can use it, too

File: input/drivers/android_input.c [hunk 1]
```diff
@@ -1398,6 +1395,13 @@ static void handle_hotplug(android_input_t *android,
    else if (strstr(android_app->current_ime, "com.hexad.bluezime"))
       strlcpy(name_buf, android_app->current_ime, sizeof(name_buf));
 
+   reserved_port = input_device_get_reserved_port(
+      vendorId, productId, device_name);
+
+   if (reserved_port > -1) {
```
Comment (LibretroAdmin) [inline input/drivers/android_input.c:1401]: Can you get rid of the brackets here? For a single line there should be no brackets

File: input/input_driver.c [hunk 2]
```diff
@@ -7289,3 +7289,61 @@ void input_keyboard_event(bool down, unsigned code,
...
+
+   int  port;
+   bool is_reserved;
+   char settings_key[256];
+   char settings_value[256];
+   int  settings_value_vendor_id;
+   int  settings_value_product_id;
+   char settings_value_device_name[256];
+   char *settings_values[MAX_USERS];
+   settings_values[0] = settings->arrays.input_player1_reserved_device;
```
Comment (LibretroAdmin) [inline input/input_driver.c:7306]: I think we should honestly dehardcode this so that we can iterate over this in the for loop below.

File: configuration.h [hunk 3]
```diff
@@ -476,6 +476,23 @@ typedef struct settings
       char input_android_physical_keyboard[255];
 #endif
 
+      char input_player1_reserved_device[255];
```
Comment (LibretroAdmin) [inline configuration.h:479]: I think instead this should be turned into a 3D array instead, and instead of a hardcoded '16 players', go with a define like MAX_USERS instead for the first array dimension.

So it'd be like

char input_players_reserved_device[MAX_USERS][255];

File: configuration.c [hunk 4]
```diff
@@ -1511,6 +1511,23 @@ static struct config_array_setting *populate_settings_array(
    SETTING_ARRAY("input_android_physical_keyboard", settings->arrays.input_android_physical_keyboard, false, NULL, true);
 #endif
 
+   SETTING_ARRAY("input_player1_reserved_device",  settings->arrays.input_player1_reserved_device,  false, NULL, true);
```
Comment (LibretroAdmin) [inline configuration.c:1514]: Same here, dehardcode all this.

File: input/input_driver.c [hunk 5]
```diff
@@ -7289,3 +7289,61 @@ void input_keyboard_event(bool down, unsigned code,
...
+
+      if (!string_is_empty(settings_value))
+      {
+         if (sscanf(settings_value, "%04x:%04x ", &settings_value_vendor_id, &settings_value_product_id) != 2)
+         {
+            strlcpy(settings_value_device_name, settings_value, sizeof(settings_value_device_name));
+            is_reserved = string_is_equal(device_name, settings_value_device_name);
+         }
+         else
+         {
```
Comment (LibretroAdmin) [inline input/input_driver.c:7335]: Single line, get rid of the brackets.

**Review by LibretroAdmin — CHANGES_REQUESTED** [review]

File: configuration.h (hunk 3 above)
Comment (PatrickStankard) [reply configuration.h:479]: Ooh, great idea 👍

File: configuration.c (hunk 4 above)
Comment (PatrickStankard) [reply configuration.c]: I was able to dehardcode everything besides these calls to `SETTING_ARRAY`, I couldn't figure out a way to get having the key dynamically generated in a loop to work, but it's probably because my C is rusty: https://github.com/libretro/RetroArch/pull/16269/commits/2aeeb237c9783a09b1902f9aa99336bf89163bbc

I noticed another player-specific setting being hardcoded so I ended up doing the same thing: https://github.com/libretro/RetroArch/blob/adaa19cbf7f5d1ad7847ab0156e052e35a20f665/configuration.c#L2141-L2156

**Comment by PatrickStankard** [comment]
@LibretroAdmin This is good for another review now 👍

**Comment by zoltanvb** [comment]
Hi @PatrickStankard,

this is one of the requests that are recurring, thank you for taking the effort!

I lifted your code and moved it to autoconfig part, and it works there as well:
https://github.com/libretro/RetroArch/compare/master...zoltanvb:RetroArch:input_driver_reserve_device

The advantage of having it in the autoconfig part would be that it is then automatically valid for all joypad drivers, Matching the driver defined ports to the player* settings is done by `input_joypad_index`. On the other hand, I did not see how android input driver calls the autoconfig, so that may be an exception.

I extended the test joypad driver and ran a few tests, looks OK so far but there may still be some corner cases, removing devices, multiple devices with same vid/pid, etc... so at least this modified version still needs work. If you agree, we could collaborate on this in some form.

**Comment by PatrickStankard** [comment]
This was merged as a part of #16647
